### PR TITLE
Workspace clone error

### LINKED ISSUE #931: Uncaught rawls errors when attempting workspace clone
Sean reported at 9:48:43 on Nov 29 2018 that he was unable to clone, and the application hung. It appears rawls returned 409 (he apparently tried to clone twice to the same name, generating conflict) and we did not handle the error. Possibly the error originated in apache and was HTML instead of JSON, causing us to misshandle it.

When we get 409, we should catch this and tell the user:
*"The workspace [name of clone] already exists. Please chose another name"*

Users are also getting 404 when attempting to clone a workspace. The reason appears to be that they are not on the authorization domain, but are on the workspace ACL. So they see the workspace in their list even though they don't really have access (see #933), and attempt to clone it from the three-dot menu on the card. We should catch this and tell the user:
"You do not appear to have access to [name of original workspace]. It is possible you are not a member of the workspace Authorization Domain. Please contact the workspace owner regarding access to this workspace."
Authorization Domain should be a link to the documentation around AD's.

The route causes of these errors will be addressed elsewhere. For now, we want to appropriately message the error to users and tell them what to do.

### PR DESCRIPTION
**Pull request by elizzhao**
Fixes #931 

409 error (workspace w/ same name already exists)
surfaces json error message on Modal in red saying "workspace [name] already exists" at the bottom.

screen shot 2018-12-07 at 2 17 30 pm

404 error (not on auth domain)
prevents this error by not giving the user permission to clone if they are not part of every group in the authorization domain of that workspace. Button is inactivated with tooltip message that says "You do not have access to the workspace authorization domain" 

screen shot 2018-12-07 at 2 18 04 pm

### CODE REVIEW

**Comment by katevoss** [comment]
@elizzhao do you have any recommendations for easily testing this out? I'm not sure how to trigger a 409 error.

File: src/pages/workspaces/List.js [hunk 1]
```diff
@@ -73,17 +74,22 @@ const styles = {
...
 const WorkspaceCard = pure(({
-  listView, onClone, onDelete, onShare,
-  workspace: { accessLevel, workspace: { namespace, name, createdBy, lastModified, attributes: { description } } }
+  listView, onClone, onDelete, onShare, groupNames,
+  workspace: { accessLevel, workspace: { namespace, name, createdBy, lastModified, authorizationDomain, attributes: { description } } }
 }) => {
   const lastChanged = `Last changed: ${Utils.makePrettyDate(lastModified)}`
   const badge = div({ title: createdBy, style: styles.badge }, [createdBy[0].toUpperCase()])
   const isOwner = Utils.isOwner(accessLevel)
+  const adGroups = _.map(({ membersGroupName }) => membersGroupName, authorizationDomain)
```
Comment (panentheos) [inline src/pages/workspaces/List.js]: FYI: could use the shorthand string form of the iterator

File: src/pages/workspaces/List.js [hunk 2]
```diff
@@ -73,17 +74,22 @@ const styles = {
...
+  const notAuthorized = groupNames && _.some(v => !groupNames.includes(v), adGroups)
```
Comment (panentheos) [inline src/pages/workspaces/List.js]: Do we need to do this ourselves, or could we utilize the `accessLevel`?

**Review by panentheos — APPROVED** [review]
Looks good, but see comment.

**Comment by andy7i** [comment]
FYI - I'm seeing this same issue when trying to create a new workspace with a duplicate namespace/name, resulting in a 409.

**Comment by elizzhao** [comment]
@katevoss the 409 error is a conflict error, so if you try and copy a workspace with the same name, it'll trigger the error. For example: you click "clone" on son-of-jimmy, and it auto-populates the field as "son-of-jimmy copy", if you delete the word "copy" so the name is now "son-of-jimmy", then it'll cause the error if you try to click the blue "clone workspace" button.

File: src/pages/workspaces/List.js (hunk 2 above)
Comment (elizzhao) [reply src/pages/workspaces/List.js]: Access level doesn't have any information about authorization domains. I don't think we currently have anything that cross-checks the groups to ensure the user is a member of all the auth domain groups necessary to access the workspace.

Comment (panentheos) [reply src/pages/workspaces/List.js]: I think the server is performing this calculation as part of `accessLevel`. I created a workspace with an auth domain, and shared it with a user that didn't have that domain. The workspace showed up in their list, but the `accessLevel` was `NO_ACCESS`. So I think we can use that as a signal for this behavior.

Comment (elizzhao) [reply src/pages/workspaces/List.js]: oh cool, so we can just use that as our check! Thanks for noticing that!

**Review by katevoss — APPROVED** [review]
Looks good! I noticed another issue with cloning that I logged in #1002 where the billing project isn't required when cloning a workspace.

**Review by breilly2 — APPROVED** [review]
👍 Looks good. So the 409 problem was just that we weren't waiting for the error JSON?

**Comment by elizzhao** [comment]
@breilly2 the 409 error before was that we weren't properly reading the message from the error json, so that part of the pr just surfaced the message properly 
@katevoss thanks for finding that and logging it!

**Pull request merged**